google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -75,68 +76,119 @@ Instant instant() {
-        ReadContext delegate, PooledSession session, boolean isSingleUse) {
-      this.delegate = delegate;
+        Supplier<T> delegateSupplier, PooledSession session, boolean isSingleUse) {
+      this.readContextDelegateSupplier = delegateSupplier;
       this.session = session;
       this.isSingleUse = isSingleUse;
+      try {
+        this.readContextDelegate = delegateSupplier.get();
+      } catch (SessionNotFoundException e) {
+        recreateSessionIfPossible(e);
+      }
     }
 
-    private ResultSet wrap(final ResultSet resultSet) {
-      session.markUsed();
-      if (!isSingleUse) {
-        return resultSet;
+    T getReadContextDelegate() {
+      return readContextDelegate;
+    }
+
+    private ResultSet wrap(final Supplier<ResultSet> resultSetSupplier) {
+      ResultSet res;
+      try {
+        res = resultSetSupplier.get();
+      } catch (SessionNotFoundException e) {
+        recreateSessionIfPossible(e);
+        res = resultSetSupplier.get();
       }
-      return new ForwardingResultSet(resultSet) {
+      return new ForwardingResultSet(res) {
+        private boolean beforeFirst = true;
+
         @Override
         public boolean next() throws SpannerException {
-          try {
-            boolean ret = super.next();
-            if (!ret) {
-              close();
-            }
-            return ret;
-          } catch (SpannerException e) {
-            if (!closed) {
-              session.lastException = e;
-              AutoClosingReadContext.this.close();
+          while (true) {
+            try {
+              boolean ret = super.next();
+              if (beforeFirst) {
+                session.markUsed();
+                beforeFirst = false;
+              }
+              if (!ret && isSingleUse) {
+                close();
+              }
+              return ret;
+            } catch (SessionNotFoundException e) {
+              recreateSessionIfPossible(e);
+              replaceDelegate(resultSetSupplier.get());
+            } catch (SpannerException e) {
+              if (!closed && isSingleUse) {
+                session.lastException = e;
+                AutoClosingReadContext.this.close();
+              }
+              throw e;
             }
-            throw e;
           }
         }
 
         @Override
         public void close() {
           super.close();
-          AutoClosingReadContext.this.close();
+          if (isSingleUse) {
+            AutoClosingReadContext.this.close();
+          }
         }
       };
     }
 
+    private void recreateSessionIfPossible(SessionNotFoundException e) {
+      session.maybeRecreateUnderlyingSession(e);
+      session.markUsed();
+      readContextDelegate = readContextDelegateSupplier.get();
+    }
+
     @Override
     public ResultSet read(
-        String table, KeySet keys, Iterable<String> columns, ReadOption... options) {
-      return wrap(delegate.read(table, keys, columns, options));
+        final String table,
```

Comment:
Can you please make style changes in a separate PR?

Reply:
This change is made automatically by the code formatter, as the method parameters are changed to final, which is needed in order to be able to reference them in the anonymous inner class.